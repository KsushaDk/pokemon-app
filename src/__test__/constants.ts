import { EvoTest, PokInfoTest } from '@utils/types';

export const group: EvoTest[] = [
  {
    chain: {
      species: {
        name: 'charmander',
      },
      evolves_to: [
        {
          species: {
            name: 'charmeleon',
          },
          evolves_to: [
            {
              species: {
                name: 'charizard',
              },
            },
          ],
        },
      ],
    },
  },
  {
    chain: {
      species: {
        name: 'bulbasaur',
      },
      evolves_to: [
        {
          species: {
            name: 'ivysaur',
          },
          evolves_to: [
            {
              species: {
                name: 'venusaur',
              },
            },
          ],
        },
      ],
    },
  },
];

export const fakePokData: PokInfoTest = {
  abilities: [
    {
      ability: {
        name: 'blaze',
        url: 'example.ability.com',
      },
      is_hidden: false,
      slot: 1,
    },
  ],
  id: 5,
  name: 'charmeleon',
  sprites: {
    other: {
      dream_world: {
        front_default: 'test.com',
        front_female: null,
      },
    },
  },
  species: {
    name: 'charmeleon',
    url: 'example.specoec.com',
  },
  types: [
    {
      slot: 1,
      type: {
        name: 'fire',
        url: 'example.type.com',
      },
    },
  ],
  stats: [
    {
      base_stat: 58,
      effort: 0,
      stat: {
        name: 'hp',
        url: 'https://pokeapi.co/api/v2/stat/1/',
      },
    },
  ],
  height: 11,
  weight: 190,
};

export const falsyPokData: PokInfoTest = {
  abilities: [
    {
      ability: {
        name: 'blaze',
        url: 'example.ability.com',
      },
      is_hidden: false,
      slot: 1,
    },
  ],
  id: 5,
  name: 'fakepok',
  sprites: {
    other: {
      dream_world: {
        front_default: 'test.com',
        front_female: null,
      },
    },
  },
  species: {
    name: 'charmeleon',
    url: 'example.specoec.com',
  },
  types: [
    {
      slot: 1,
      type: {
        name: 'fire',
        url: 'example.type.com',
      },
    },
  ],
  stats: [
    {
      base_stat: 58,
      effort: 0,
      stat: {
        name: 'hp',
        url: 'https://pokeapi.co/api/v2/stat/1/',
      },
    },
  ],
  height: 11,
  weight: 190,
};

export const groupResult: EvoTest = {
  chain: {
    species: {
      name: 'charmander',
    },
    evolves_to: [
      {
        species: {
          name: 'charmeleon',
        },
        evolves_to: [
          {
            species: {
              name: 'charizard',
            },
          },
        ],
      },
    ],
  },
};

export const fakeGroupResult: EvoTest = {
  chain: {
    species: {
      name: 'charmander',
    },
    evolves_to: [
      {
        species: {
          name: undefined,
        },
        evolves_to: [
          {
            species: {
              name: undefined,
            },
          },
        ],
      },
    ],
  },
};

export const allFakeUrls: string[] = [
  'https://pokeapi.co/api/v2/pokemon/charmander',
  'https://pokeapi.co/api/v2/pokemon/charmeleon',
  'https://pokeapi.co/api/v2/pokemon/charizard',
];

export const fakeUrls: string[] = [
  'https://pokeapi.co/api/v2/pokemon/charmander',
];
