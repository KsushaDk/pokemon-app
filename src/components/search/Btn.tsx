import React, { FC } from 'react';

type BtnProps = {
  btnValue: string;
  onClick: React.MouseEventHandler<HTMLButtonElement>;
};

export const Btn: FC<BtnProps> = (props) => {
  const { btnValue, onClick } = props;

  return (
    <button className="btn" type="button" onClick={onClick}>
      {btnValue}
    </button>
  );
};
