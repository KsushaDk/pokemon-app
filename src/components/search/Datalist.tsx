import React, { FC } from 'react';
import { useDispatch } from 'react-redux';

import { setPokForSearch, setTypeForSearch } from '@redux/actions/actions';

type DatalistProps = {
  datainfo: string[];
  id: string;
};

export const Datalist: FC<DatalistProps> = ({ datainfo, id }) => {
  const dispatch = useDispatch();

  const handleChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    if (id === 'type') {
      dispatch(setTypeForSearch(e.target.options[e.target.selectedIndex].text));
    } else {
      dispatch(setPokForSearch(e.target.options[e.target.selectedIndex].text));
    }
  };

  return (
    <select className="select" id={id} onChange={handleChange}>
      {datainfo.map((item) => (
        <option value={item} key={item}>
          {item}
        </option>
      ))}
    </select>
  );
};
