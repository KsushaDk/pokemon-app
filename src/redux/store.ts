import { createStore, compose, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';

import { pokemonsReducers } from './reducers/pokemonsReducers';
import { urlsSaga } from './sages/urls';

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
  }
}

const sagaMiddleware = createSagaMiddleware();

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const store = createStore(
  pokemonsReducers,
  applyMiddleware(sagaMiddleware)
  // composeEnhancers()
);

sagaMiddleware.run(urlsSaga);
