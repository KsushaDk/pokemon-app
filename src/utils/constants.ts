export const URL_FOR_POK_LIST =
  'https://pokeapi.co/api/v2/pokemon/?offset=0&limit=20';
export const URL_FOR_EVO_LIST =
  'https://pokeapi.co/api/v2/evolution-chain/?offset=0&limit=20';
