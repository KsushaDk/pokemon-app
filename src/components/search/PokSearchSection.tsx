import React, { FC } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';

import { setLoading, setSearchedPokData } from '@redux/actions/actions';
import { IPoksState } from '@redux/reducers/pokemonsReducers';
import { httpGet } from '@utils/request';

import { Btn } from './Btn';
import { Datalist } from './Datalist';

export const PokSearchSection: FC = () => {
  const dispatch = useDispatch();

  const pokUrls = useSelector<IPoksState, IPoksState['pokUrls']>(
    (state) => state.pokUrls
  );

  const nameArr = pokUrls.map((item) => item.name);

  const pokForSearch = useSelector<IPoksState, IPoksState['pokForSearch']>(
    (state) => state.pokForSearch
  );

  const handleClick: React.MouseEventHandler<HTMLButtonElement> = () => {
    if (pokForSearch === '') {
      alert('Please select pokemon');
    }
    dispatch(setLoading(true));
    const searchedPokData = httpGet(
      `https://pokeapi.co/api/v2/pokemon/${pokForSearch}`
    );
    searchedPokData.then((result) => {
      dispatch(setSearchedPokData(result));
      dispatch(setLoading(false));
    });
  };

  return (
    <>
      <h5 className="header__title">Search pokemon</h5>
      <form className="header__form">
        <Datalist datainfo={nameArr} id="character" />

        <Link to={pokForSearch ? '/search' : '#'}>
          <Btn btnValue="Search" onClick={handleClick} />
        </Link>
      </form>
    </>
  );
};
