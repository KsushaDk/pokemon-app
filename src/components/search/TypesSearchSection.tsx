import React, { FC } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';

import { Btn } from './Btn';
import { Datalist } from './Datalist';

import { setSearchedTypesGroup } from '@redux/actions/actions';
import { IPoksState } from '@redux/reducers/pokemonsReducers';
import { POK_TYPES } from '@utils/constants';

export const TypesSearchSection: FC = () => {
  const dispatch = useDispatch();

  const pokData = useSelector<IPoksState, IPoksState['pokData']>(
    (state) => state.pokData
  );

  const typeForSearch = useSelector<IPoksState, IPoksState['typeForSearch']>(
    (state) => state.typeForSearch
  );

  const getFilteredArr = () => {
    if (typeForSearch === '') {
      alert('Please select pokemon type');
      return null;
    }
    return pokData.filter((pok) => {
      const types = pok.types.map((item) => item.type.name);
      if (types.includes(typeForSearch)) {
        return pok;
      }
    });
  };

  const handleClick: React.MouseEventHandler<HTMLButtonElement> = () => {
    const filteredArr = getFilteredArr();
    if (filteredArr !== null) {
      dispatch(setSearchedTypesGroup(filteredArr));
    }
  };

  return (
    <>
      <h5 className="header__title">Filter Pokemons by Type:</h5>
      <form className="header__form">
        <Datalist datainfo={POK_TYPES} id="type" />
        <Link to={typeForSearch ? '/types' : '#'}>
          <Btn btnValue="Search" onClick={handleClick} />
        </Link>
      </form>
    </>
  );
};
