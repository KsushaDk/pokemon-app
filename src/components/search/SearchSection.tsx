import React, { FC } from 'react';

import { PokSearchSection } from './PokSearchSection';
import { TypesSearchSection } from './TypesSearchSection';

export const SearchSection: FC = () => {
  return (
    <header className="header">
      <PokSearchSection />

      <TypesSearchSection />
    </header>
  );
};
