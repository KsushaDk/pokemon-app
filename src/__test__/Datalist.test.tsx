import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Datalist } from '@components/search/Datalist';
import { POK_TYPES } from '@utils/constants';
import { store } from '@redux/store';

const data = POK_TYPES;

describe('Datalist component', () => {
  let datalist: any;

  beforeEach(() => {
    datalist = render(
      <Provider store={store}>
        <Datalist datainfo={data} id={'type'} />
      </Provider>
    );
  });

  test('Datalist renders', () => {
    expect(screen.getByText(/ground/)).toBeInTheDocument();
  });

  test('Datalist shapshot', () => {
    expect(datalist).toMatchSnapshot();
  });
});
