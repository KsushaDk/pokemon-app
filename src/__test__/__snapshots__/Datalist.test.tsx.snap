// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datalist component Datalist shapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <select
        class="select"
        id="type"
      >
        <option
          value="ground"
        >
          ground
        </option>
        <option
          value="poison"
        >
          poison
        </option>
        <option
          value="normal"
        >
          normal
        </option>
        <option
          value="flying"
        >
          flying
        </option>
        <option
          value="rock"
        >
          rock
        </option>
        <option
          value="bug"
        >
          bug
        </option>
        <option
          value="fighting"
        >
          fighting
        </option>
        <option
          value="ghost"
        >
          ghost
        </option>
        <option
          value="steel"
        >
          steel
        </option>
        <option
          value="fire"
        >
          fire
        </option>
        <option
          value="water"
        >
          water
        </option>
        <option
          value="grass"
        >
          grass
        </option>
        <option
          value="electric"
        >
          electric
        </option>
        <option
          value="pyschic"
        >
          pyschic
        </option>
        <option
          value="ice"
        >
          ice
        </option>
        <option
          value="dragon"
        >
          dragon
        </option>
        <option
          value="dark"
        >
          dark
        </option>
        <option
          value="fairy"
        >
          fairy
        </option>
      </select>
    </div>
  </body>,
  "container": <div>
    <select
      class="select"
      id="type"
    >
      <option
        value="ground"
      >
        ground
      </option>
      <option
        value="poison"
      >
        poison
      </option>
      <option
        value="normal"
      >
        normal
      </option>
      <option
        value="flying"
      >
        flying
      </option>
      <option
        value="rock"
      >
        rock
      </option>
      <option
        value="bug"
      >
        bug
      </option>
      <option
        value="fighting"
      >
        fighting
      </option>
      <option
        value="ghost"
      >
        ghost
      </option>
      <option
        value="steel"
      >
        steel
      </option>
      <option
        value="fire"
      >
        fire
      </option>
      <option
        value="water"
      >
        water
      </option>
      <option
        value="grass"
      >
        grass
      </option>
      <option
        value="electric"
      >
        electric
      </option>
      <option
        value="pyschic"
      >
        pyschic
      </option>
      <option
        value="ice"
      >
        ice
      </option>
      <option
        value="dragon"
      >
        dragon
      </option>
      <option
        value="dark"
      >
        dark
      </option>
      <option
        value="fairy"
      >
        fairy
      </option>
    </select>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
